staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -599,6 +607,34 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
 		s.NonLongRunningRequestWaitGroup.Wait()
 	}()
 
+	// wait for all in-flight watches to finish
+	activeWatchesDrainedCh := make(chan struct{})
+	go func() {
+		defer close(activeWatchesDrainedCh)
+
+		<-notAcceptingNewRequestCh.Signaled()
+
+		// Wait for all active watches to finish
+		grace := 10 * time.Second
+		activeBefore, activeAfter, err := s.WatchRequestWaitGroup.Wait(func(count int) (utilwaitgroup.RateLimiter, context.Context, context.CancelFunc) {
+			ctx, cancel := context.WithTimeout(context.Background(), grace)
+			// TODO: a) do we need to make this grace period configurable?
+			//  b) we need to tweak the qps to be more robust.
```

Comment:
I would suggest going with something like `max(count/grace.seconds(), <some constant>)`

Basically what I would like to achieve is that when the number of watches is relatively small, we don't need to wait for the whole grace-period, as shutting them down faster is safe enough.

Reply:
i have added a back of the envelope for now, we can tweak more later